File: Python/problems/3637.py
from typing import List


class Solution:
    def _getInd(self, nums, cur_ind, increasing = True):
        N = len(nums)
        if cur_ind == N:
            return N

        for i in range(cur_ind, N-1):
            if increasing and nums[i] >= nums[i + 1]:
                return i
            elif not increasing and nums[i] <= nums[i+1]:
                return i

        return N
                
    def isTrionic(self, nums: List[int]) -> bool:
        N = len(nums)
        p = self._getInd(nums, 0)
        if p >= N-2 or p == 0:
            return False
        q = self._getInd(nums, p, increasing = False)
        if q >= N-1:
            return False
        r = self._getInd(nums, q)
        return r == N

File: Rust/3637.rs
impl Solution {
    pub fn is_trionic(nums: Vec<i32>) -> bool {
        let n = nums.len();
        let p = Self::get_ind(&nums, 0, true);
        if p>=n-2 || p == 0 {
            return false;
        }
        let q = Self::get_ind(&nums, p, false);
        if q >= n-1 {
            return false;
        }
        Self::get_ind(&nums, q, true) == n
    }

    pub fn get_ind(nums: &Vec<i32>, cur_ind: usize, increasing: bool) -> usize {
        let n = nums.len();
        for i in cur_ind..n-1 {
            if (increasing && nums[i] >= nums[i+1]) || (!increasing && nums[i] <= nums[i+1]) {
                return i;
            }
        }
        n
    }
}